python/pyarrow/array.pxi

```diff
@@ -524,6 +528,20 @@ cdef class Array:
                                               self, &out))
         return wrap_array_output(out)
 
+    def to_numpy(self):
+        """
+        Convert to a NumPy array
+        """
+        if self.null_count:
+            raise NotImplementedError('NumPy array conversion is only '
+                                      'supported for arrays without nulls.')
+        if not is_primitive(self.type.id):
+            raise NotImplementedError('NumPy array conversion is only '
+                                      'supported for primitive types.')
+        buflist = self.buffers()
+        return np.frombuffer(buflist[-1], dtype=self.type.to_pandas_dtype())[
+            self.offset:self.offset + len(self)]
```

Comment:
This feels like a temporary workaround. If we accept this patch, can you open a JIRA about implementing this in libarrow_python (i.e. in C++) where we can utilize common code paths with `to_pandas` and toggle between NumPy-for-pandas and NumPy-for-NumPy behavior (and use the zero_copy_only flag where needed)?